### DISCUSSION TITLE
[BP#53] Éviter les boucles for...in

### DISCUSSION THREAD

**Discussion by didierfred** [cnumr/best-practices#81] [Bonnes pratiques] [answered]
La boucle foreach est tout a fait acceptable à mon sens 

Ci joint un [script de test](https://github.com/cnumr/best-practices/files/7405219/test_forin.zip) , on voit clairement la différence pour le boucle for in mais pour foreach vs for c'est insiginifiant sur mon PC (1 ms d'ecart pour une iteration d'un tableau d'un million d'entrée)

**Comment by devcrafting** [comment]
Probablement à tester sur d'autres languages, mais cela ne me semble en effet pas une pratique bien utile d'un point de vue sobriété, sans compter qu'elle a un impact sur la lisibilité/maintenabilité

**Comment by ACTLEM** [comment, marked as answer]
https://github.com/cnumr/best-practices/issues/143

**Reply by devcrafting** [reply]
Pas sûr de comprendre ce qu'il faut modifier ? Les 2 commentaires allaient plutôt dans le sens de la suppression non? Et spécifique à un language ?

**Reply by ACTLEM** [reply]
La BP est conservée par rapport à la différence assez criante entre le for/foreach et le for...in (facteur de 8 à 10 pour des tests réalisés pourtant sur des "petits" (200 éléments) tableaux, encore plus flagrant sur des grands tableaux (passage de 3ms à 500ms sur le test de Frédéric). Par contre le foreach ne change pas grand chose donc il faut supprimer la référence à cette instruction.

**Reply by tbroyer** [reply]
`for…in` est déconseillé pour les itérables (source de bugs), il y a `for…of` pour ça.

* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

On tombe donc dans le bon usage du langage, des bonnes pratiques de développement plus que d'écoconception finalement.

Sur les objets (non itérables), il faudrait comparer `for…in`+`hasOwnProperty` (cf. https://eslint.org/docs/rules/guard-for-in par exemple) avec une itération sur `Object.keys()`, voire une `Map` (spoiler: le résultat dépendra des usages: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps) ; à moins qu'on veuille réellement boucler sur toutes les propriétés y compris celles héritées par la chaîne des prototypes et à ce moment là la question ne se pose pas vraiment (`Object.keys()` et une récursion sur `Object.getPrototypeOf()` ? :thinking:).

**Reply by devcrafting** [reply]
Du coup c'est spécifique JS => à mettre dans le titre "en JS"
